### DISCUSSION TITLE
How to apply transparent background to iframe?

### DISCUSSION THREAD

**Discussion by TetraTheta** [hbstack/theme#393] [Q&A] [answered]
I want to embed Steam Widget to my post with this `<iframe>` code which Steam provided:
```html
<iframe src="https://store.steampowered.com/widget/244910/" frameborder="0" width="646" height="190"></iframe>
```

Since `background` is set to `transparent` in `<body>` tag of the iframe content, it should be displayed like this:

![IMG1695817590](https://github.com/hbstack/theme/assets/37061801/1f3a0212-8c0a-48c2-af72-0e6b653a8c72)

But instead, I get this white background.

![IMG1695817361](https://github.com/hbstack/theme/assets/37061801/f30b692b-d254-419a-b486-9e7a4cf1cafa)

This is not the problem of Steam Widget. I've used same HTML code at other site, and its background was transparent (showing site's background color).

Adding `allowtransparency="true"` to `<iframe>` didn't help.

How can I solve this problem?

**Comment by razonyang** [comment]
It works fine on light mode, and the issue happens on refreshing the pages on `dark` mode, but I don't have an ideal solution.

**Reply by TetraTheta** [reply]
My first guess was something is setting iframe's background color to white, but I couldn't find any style or attribute that does that on my web browser's inspector.

In theory, if nothing is modifying background of iframe, its `transparent` should be respected...

**Reply by razonyang** [reply]
I don't know how to fix, just tested in Google dark mode (`F12` and put the iframe code into the HTML), the iframe doesn't work either, _change the iframe backgroud (`style="background: red;"`) affects on light mode only_. Please feel free to share docs if you found any.

![image](https://github.com/hbstack/theme/assets/17720932/01e9b66d-6514-4c23-bc92-0cdf249566ce)

**Reply by TetraTheta** [reply]
I think only viable workaround is mimic Steam Widget's layout using shortcode.

Fortunately, Steam provides API for it: <https://store.steampowered.com/api/appdetails?appids=244910>. I can grab `result.data.name`, `result.data.short_description`, `result.data.price_overview.final_formatted`, `result.data.platforms` to get required data.

Since I can't modify content of iframe because of SOP(Same-Origin Policy), this seems the only option for me.

The situation has jumped out of the frying pan into the fire, but it seems this is only usable option 😂

**Reply by TetraTheta** [reply, marked as answer]
Well, it turned out that I can't fetch JSON from Steam. CORS or SOP or whatever the 'security stuff' blocks web browser to fetch it from Steam because client (web browser) is not on same domain of Steam. 

If then, I will just give up using iframe and shortcode. This thing is not that critical stuff to me, so I will just use 'link preview'-like shortcode for it.

Thanks for the help!

**Comment by razonyang** [comment]
You can also use Hugo [getJSON](https://gohugo.io/templates/data-templates/#get-remote-data) to fetch API, there won't a CORS issues (I guess you're using JS?).

```html
// layouts/shortcodes/steam.html
{{- with getJSON "https://store.steampowered.com/api/appdetails?appids={{ .Get 0 }}" }}
  {{- $data := dict }}
  {{- range . }}
    {{- $data = .data }}
    {{- break }}
  {{- end }}
  <div class="bg-dark-subtle p-3 mb-3">
    <p class="h5">{{ $data.name }}</p>
    <div class="d-flex">
      <div class="flex-shrink-0">
        <img src="{{ $data.capsule_image }}" alt="{{ $data.name }}">
      </div>
      <div class="flex-grow-1 ms-3">
        {{ $data.short_description | safeHTML }}
      </div>
    </div>
  </div>
{{- end }}
```

And then use the shortcode in content.

```markdown
{{< steam "244910" >}}
```

Please note that, the API may has limitations, such as authentication, rate limit and number of requests, you can find those info on their docs and Hugo docs.
